File: python/test/test_HODLR.py
import numpy as np
import pyhodlrlib

# Size of the matrix:
N = 500
x = np.sort(np.random.rand(N))

# Returning the Gaussian Kernel:
class Kernel(pyhodlrlib.HODLR_Matrix):
    def getMatrixEntry(self, i, j):
        if(i == j):
            return 10
        else:
            return np.exp(-(x[i] - x[j])**2)

def test_functions(factorization_type, is_sym, is_pd):
    K = Kernel(N)
    # What we are doing here is explicitly generating 
    # the matrix from its entries
    A = K.getMatrix(0, 0, N, N)
    # Tolerance for all factorizations:
    eps = 1e-12
    # Declaring the Factorizer Object:
    F = pyhodlrlib.Matrix_Factorizer(K, factorization_type)

    # Size of leaf level:
    M = 50
    # Number of levels:
    n_levels = int(np.log(N / M) / np.log(2))

    # Creating the HODLR Tree object:
    T = pyhodlrlib.HODLR_Tree(n_levels, eps, F)
    T.assembleTree(is_sym, is_pd)

    # Random vector to take product with:
    x = np.random.rand(N)
    # Finding b using HODLR:
    b_hodlr = T.matmatProduct(x)
    # Finding b using direct MatVec:
    b = A @ x
    # Verifying the accuracy of the MatMat:
    assert(np.linalg.norm(b_hodlr.ravel() - b.ravel()) / np.linalg.norm(b) < N * eps)

    # Factorize elements of the tree:
    T.factorize()
    # Solving for x in A x = b:
    x_hodlr = T.solve(b)
    # Computing the relative error:
    assert(np.linalg.norm(x_hodlr.ravel() - x) /  np.linalg.norm(x) < N * eps)

    # Finding log determinant:
    logdet_hodlr = T.logDeterminant()
    # Finding logdet:
    logdet = 2 * np.sum(np.log(abs(np.diag(np.linalg.cholesky(A)))))
    assert(abs(logdet_hodlr - logdet) / logdet < N * eps)

    # When system described is SPD:
    if(is_sym and is_pd):
        # Setting y = W^T x
        y = T.symmetricFactorTransposeProduct(x)
        # Getting b = W (W^T x)
        b_hodlr = T.symmetricFactorProduct(y)
        assert(np.linalg.norm(b_hodlr.ravel() - b.ravel()) / np.linalg.norm(b) <  N * eps)

        # Directly obtaining the symmetric factor matrix:
        W = T.getSymmetricFactor()
        assert(np.mean(abs(W @ W.T - A)) < N * eps)

    return

test_functions('rookPivoting', False, False)
test_functions('rookPivoting', True, False)
test_functions('rookPivoting', True, True)

test_functions('queenPivoting', False, False)
test_functions('queenPivoting', True, False)
test_functions('queenPivoting', True, True)

test_functions('SVD', False, False)
test_functions('SVD', True, False)
test_functions('SVD', True, True)

test_functions('RRQR', False, False)
test_functions('RRQR', True, False)
test_functions('RRQR', True, True)

print('Reached End of Test File Successfully! All functions work as intended!')

File: fortran/test/test_HODLR.f90
! Function that returns the matrix values:
function get_matrix_entry(i, j) bind(c, name="get_matrix_entry")
    implicit none
    integer, intent(in) :: i
    integer, intent(in) :: j

    ! It needs to be ensured that this is the same as N set below:
    integer, parameter :: N = 1000
    double precision :: get_matrix_entry

    if(i == j) then
        get_matrix_entry = 100
    else
        get_matrix_entry = exp(-(DBLE(i)/N - DBLE(j)/N)**2)
    end if

end function

program main

    use iso_c_binding
    use hodlr_mod

    implicit none

    ! The following objects are used to refer to the C++ objects
    type(c_ptr) :: kernel
    type(c_ptr) :: hodlr

    ! Size of the matrix:
    integer, parameter :: N = 1000
    ! Size of the matrix at leaf level:
    integer, parameter :: M = 200
    ! Number of digits of tolerance required:
    real(c_double), parameter :: eps = 1.0d-15

    ! Whether matrix encoded is symmetric:
    logical(c_bool) :: is_sym = .true.
    ! Whether matrix encoded is positive-definite:
    logical(c_bool) :: is_pd = .true.

    ! Used to store the complete matrix:
    real(c_double) :: flattened_matrix(N * N)
    real(c_double) :: matrix(N, N)

    ! Variables used to store the symmetric factor matrix:
    real(c_double) :: flattened_W_matrix(N * N)
    real(c_double) :: W_matrix(N, N)

    integer :: i, j

    ! Creating random vectors used on the RHS and LHS for multiplication and solving:
    real(c_double) :: x(N)
    real(c_double) :: b(N)
    real(c_double) :: b_exact(N)
    real(c_double) :: x_approx(N)
    real(c_double) :: y(N)

    ! Used to store the log determinant of the matrix:
    real(c_double) :: log_det
    ! Log-Det exact for this matrix with N = 1000:
    real(c_double) :: log_det_exact = 4598.3043670568

    ! Method used for low-rank matrix decomposition:
    character(len = 20) :: factorization_method = "rookPivoting"

    ! Creating the kernel object:
    call initialize_kernel_object(N, kernel)

    ! Getting the matrix encoded by the kernel object:
    call get_matrix(kernel, 0, 0, N, N, flattened_matrix)
    matrix = reshape(flattened_matrix, (/ N, N /))

    ! Building the HODLR object:
    call initialize_hodlr(N, M, eps, hodlr)
    call assemble(hodlr, kernel, "rookPivoting", is_sym, is_pd)

    ! Populating the vector x with random elements:
    do i = 1, N
        call random_number(x(i)) 
    end do
    
    ! Performing matrix multiplication using HODLR:
    call matmat_product(hodlr, x, b)
    ! Performing the matrix multiplication directly:
    b_exact = matmul(matrix, x)
    ! Checking the error of matmul:
    if((sum(abs(b_exact - b)) / sum(abs(b_exact)) > N * eps)) then
        print *, "POSSIBLE BUG!!: Large error in MatMul"
        call exit(1)
    end if

    ! Factorizing the elements of the tree:
    call factorize(hodlr)

    ! Solve for x in the equation Ax = b:
    call solve(hodlr, b_exact, x_approx)

    ! Checking the error of solve:
    if((sum((x_approx - x)**2) / sum(x**2) > N * eps)) then
        print *, "POSSIBLE BUG!!: Large error in Solve(Forward)"
        call exit(1)
    end if

    if((sum((matmul(matrix, x) - b_exact)**2) / sum(b_exact**2) > N * eps)) then
        print *, "POSSIBLE BUG!!: Large error in Solve(Backward)"
        call exit(1)
    end if

    ! Getting the determinant of the matrix encoded:
    call logdeterminant(hodlr, log_det)
    if(abs(log_det - log_det_exact) / log_det_exact > 1e-7) then
        print *, "POSSIBLE BUG!!: Large error in log determinant"
        call exit(1)
    end if

    ! We set y = W^T x
    call symmetric_factor_transpose_product(hodlr, x, y)
    ! b = W y = W W^T x = B * x
    call symmetric_factor_product(hodlr, y, b)
    if(sum((b_exact - b)**2) / sum(b_exact**2) > N * eps) then
        print *, "POSSIBLE BUG!!: Large error in symmetric factor products"
        call exit(1)
    end if

    ! Getting the matrix encoded by the kernel object:
    call get_symmetric_factor(hodlr, flattened_W_matrix)
    W_matrix = reshape(flattened_W_matrix, (/ N, N /))
    if(sum(abs(matrix - matmul(W_matrix, transpose(W_matrix)))) / sum(matrix) > N * eps) then
        print *, "POSSIBLE BUG!!: Large error in obtaining symmetric factor"
        call exit(1)
    end if

    print *, "Reached End of Test File Successfully! All functions work as intended!"
end
